src/metrics/abc.rs

```diff
@@ -0,0 +1,803 @@
+            .sqrt()
+    }
+
+    #[inline(always)]
+    pub(crate) fn compute_sum(&mut self) {
+        self.assignments_sum += self.assignments;
+        self.branches_sum += self.branches;
+        self.conditions_sum += self.conditions;
+    }
+
+    #[inline(always)]
+    pub(crate) fn compute_minmax(&mut self) {
+        self.assignments_min = self.assignments_min.min(self.assignments);
+        self.assignments_max = self.assignments_max.max(self.assignments);
+        self.branches_min = self.branches_min.min(self.branches);
+        self.branches_max = self.branches_max.max(self.branches);
+        self.conditions_min = self.conditions_min.min(self.conditions);
+        self.conditions_max = self.conditions_max.max(self.conditions);
+        self.compute_sum();
+    }
+}
+
+#[doc(hidden)]
+pub trait Abc
+where
+    Self: Checker,
+{
+    fn compute(_node: &Node, _stats: &mut Stats) {}
+}
+
+/// Counts Java unary conditional expressions by exploring parenthesis and not operators.
+/// If successful, it returns an Option containing a pair of values representing:
+/// - The unary condition count.
+/// - The presence of a not operator.
+///
+/// Returns None if it fails to explore deeper.
+///
+/// "A unary conditional expression is an implicit condition that uses no relational operators."
+///
+/// - Jerry Fitzpatrick (1997) - "Applying the ABC Metric to C, C++, and Java":
+/// https://www.softwarerenovation.com/Articles.aspx (page 5)
+///
+/// "These are cases where a single variable/field/value is treated as a boolean value.
+/// Examples include if (x) and return !ready."
+///
+/// - GMetrics (a Groovy source code analyzer):
+/// https://dx42.github.io/gmetrics/metrics/AbcMetric.html
+/// - GMetrics test examples:
+/// https://github.com/dx42/gmetrics/blob/master/src/test/groovy/org/gmetrics/metric/abc/AbcMetric_MethodTest.groovy
+fn count_java_unary_conditions(node: &Node) -> Option<(f64, bool)> {
+    use Java::*;
+
+    let mut res = 0.;
+    let mut condition_node;
+    let mut condition_kind;
+    let mut cursor = node.object().walk();
+    let mut is_bool = false;
+
+    // Scans the children nodes of the argument node
+    if cursor.goto_first_child() {
+        loop {
+            condition_node = cursor.node();
+            condition_kind = condition_node.kind_id().into();
+
+            // Counts unary conditions
+            if matches!(condition_kind, MethodInvocation | Identifier | True | False) {
+                res += 1.;
+            }
+
+            // Ignores parenthesis and not operators and looks inside them
+            while let ParenthesizedExpression | UnaryExpression = condition_kind {
+                // Sets flag if a not operator is found
+                // This is useful to know if a variable or a value returned by a method is proved to be a boolean
+                // e.g. `return (!x);`
+                if matches!(condition_kind, UnaryExpression) {
+                    is_bool = true;
```

Comment:
We can replace the `if` with: `is_bool = matches!(condition_kind, UnaryExpression)`